tests/integrationv2/utils.py

```diff
@@ -47,15 +48,33 @@ def invalid_test_parameters(*args, **kwargs):
     cipher = kwargs.get('cipher')
     curve = kwargs.get('curve')
 
+    if protocol is Protocols.TLS13 and get_flag(S2N_OPENSSL_VERSION) == "openssl-1.0.2-fips":
+        return True
+
     if cipher is not None:
         # If the selected protocol doesn't allow the cipher, don't test
-        if protocol is not None and cipher.min_version > protocol:
-            return True
-
-        # NOTE: We don't detect the version of OpenSSL at the moment,
-        # so we will deselect these tests.
-        if provider is not None and provider is OpenSSL and cipher.openssl1_1_1 is False:
-            return True
+        if protocol is not None:
+            if cipher.min_version > protocol:
+                return True
+            if protocol is Protocols.TLS13 and cipher.min_version < protocol:
+                return True
+
+        if provider is OpenSSL:
+            if get_flag(S2N_OPENSSL_VERSION) == "openssl-1.1.1" and cipher.openssl1_1_1 is False:
+                # If the cipher is not supported in OpenSSL 1.1.1, deselect the test
+                return True
+            if get_flag(S2N_OPENSSL_VERSION) == "openssl-1.0.2-fips":
+                if cipher.fips is False:
+                    # If the cipher is not supported using FIPS, deselect the test
+                    return True
+                invalid_ciphers = [
+                    Ciphers.ECDHE_RSA_AES128_SHA256,
+                    Ciphers.ECDHE_RSA_AES256_SHA384,
+                    Ciphers.ECDHE_RSA_AES128_GCM_SHA256,
+                    Ciphers.ECDHE_RSA_AES256_GCM_SHA384
+                ]
+                if curve is Curves.P384 and cipher in invalid_ciphers:
+                    return True
```

Comment:
Minimally this probably needs a comment. Why can't we use P384 with these specific ciphers only in fips?

Reply:
You can read the OpenSSL wiki here: https://wiki.openssl.org/index.php/FIPS_mode_and_TLS.
It is, quoting the wiki, "a bloody mess". I don't think we can capture this in the comments, other than to say some combinations are not supported in FIPS (which is exactly how ciphers do this in the existing framework., They have a boolean flag indicating fips support). I'm going through more combinations, but this is not going to be perfect.